go/vt/sqlparser/token.go

```diff
@@ -427,13 +427,24 @@ func KeywordString(id int) string {
 // Lex returns the next token form the Tokenizer.
 // This function is used by go yacc.
 func (tkn *Tokenizer) Lex(lval *yySymType) int {
+	if tkn.SkipToEnd {
+		return tkn.skipStatement()
+	}
+
 	typ, val := tkn.Scan()
 	for typ == COMMENT {
 		if tkn.AllowComments {
 			break
 		}
 		typ, val = tkn.Scan()
 	}
+	if typ == 0 || typ == ';' || typ == LEX_ERROR {
+		// If encounter end of statement or invalid token,
+		// we should not accept partially parsed DDLs. They
+		// should instead result in parser errors. See the
+		// Parse function to see how this is handled.
+		tkn.partialDDL = nil
```

Comment:
Without this reset, a statement like this would get quietly accepted by vitess: `incomplete ddl; evil query`.
This is the case where a `;` tirggers a parser error. That will cause the `Parse` function to detect that it was a partial DDL. So, it will accept the entire string, including `evil query`. It's the second test case in `TestSkipToEnd`.

Reply:
I guess I'm confused about the specific sequence we expect to happen in that evil example above.
From my read of the code, it flows like this:

1. The incomplete DDL sets `SkipToEnd`
2. The next time through the lexer it should scan ahead to the `;` character using `skipStatement`, but no further
3. It then returns to `Parse` with the partialDDL flag set.
4. Parse then calls `Scan` again expecting only a 0 but finds more tokens from `evil query` and fails then.

Presumably I'm missing something but can you clarify where my logic is wrong?

Reply:
After some DM discussion in Slack I now understand.